pkg/segment/query/processor/dataprocessor.go

```diff
@@ -107,6 +115,10 @@ func (dp *DataProcessor) Fetch() (*iqr.IQR, error) {
 				return nil, utils.TeeErrorf("DP.Fetch: failed to fetch input: %v", err)
 			}
 
+			if dp.isCleanupCalled {
+				return nil, io.EOF
+			}
+
 			dp.processorLock.Lock()
 			output, err = dp.processor.Process(input)
```

Comment:
Once you have the lock, you have to check `isCleanupCalled` again (it may have changed while we were getting the lock). So move the check to just above calling Process(). The check near the top of this function (before getting stream input) is fine as is without being locked